[NativeAOT-LLVM] More miscellaneous fixes

SingleAccretion opened the pull request:
Fixing various bugs and improving the documentation:

1) Disable background GC.
2) Add in some missed `pthread` shims.
3) Add various linker settings for stripping debug information.
4) Specify the global base and stack size.
5) Add notes about `NativeDebugSymbols` and `UnmanagedEntryPointsAssembly` to the documentation.
6) Remove `HFIF_NO_RPI_OR_GC` from the RPI helper.

SingleAccretion committed:
Disable background GC support

Turns out it was enabled all along!

SingleAccretion committed:
Stub out more threading

SingleAccretion committed:
Add some notes about exports/imports

SingleAccretion committed:
A couple more doc updates

SingleAccretion committed:
Add some linker options

1) "--strip-all" so that we strip both DWARF and the names section
   if the user specifies "/p:NativeDebugSymbols=false".
2) Also compress relocations if we are throwing away the debug info.
   Recall this saves ~8% in terms of code size.
2) Fix the global base and stack size.

SingleAccretion committed:
Add a note about NativeDebugSymbols to the doc

RReverser commented on src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.targets:
```diff
@@ -510,6 +512,8 @@ The .NET Foundation licenses this file to you under the MIT license.
       <CustomLinkerArg Condition="'$(NativeDebugSymbols)' == 'true'" Include="-g3" />
+      <CustomLinkerArg Condition="'$(NativeDebugSymbols)' != 'true'" Include="-Wl,--strip-all" />
+      <CustomLinkerArg Condition="'$(NativeDebugSymbols)' != 'true'" Include="-Wl,--compress-relocations" />
       <CustomLinkerArg Condition="'$(Optimize)' != 'true'" Include="$(WasmOptimizationSetting) -flto" />
```
Random question, not related to this PR but noticed because it was nearby: why is LTO enabled and `WasmOptimizationSetting` passed to the linker only when `$(Optimize) != 'true'`? Shouldn't it be the opposite?

SingleAccretion replied:
> why is LTO enabled

I believe that's more or less a leftover from a previous time (it used to be that the linker was passed bitcode files directly without the `.o` step). It is not used currently.

> Shouldn't it be the opposite?

It is a bug, yes, but it's not abundantly clear what the right setting actually is. See #2357.

SingleAccretion commented:
The GC test still fails, now with heap corruption. Will take a while to debug, most likely.

Edit: turned out to be a GC hole due to a codegen bug.

SingleAccretion committed:
Do not designate the RPI helper as HFIF_NO_RPI_OR_GC

The helper may invoke runtime initialization, which runs managed code.
Not saving the shadow stack beforehand can lead to various kinds of
hard-to-debug issues. For example, one failure mode was that we would
try to allocate two shadow stacks, leading later to a GC hole because
the shadow stack bottom would be above its top.

SingleAccretion commented:
@dotnet/nativeaot-llvm

yowl reviewed (approved):
Thanks for adding the memory layout options.

jkotas commented on src/coreclr/gc/gcpriv.h:
```diff
@@ -184,6 +184,11 @@ inline void FATAL_GC_ERROR()
 
 #define BACKGROUND_GC   //concurrent background GC (requires WRITE_WATCH)
 
+// It is no longer possible to build without BACKGROUND_GC defined, so we have to use this duplicate
```
@cshung You have been fixing the build without `BACKGROUND_GC` defined in the past: https://github.com/dotnet/runtime/pull/77328 . Is building without `BACKGROUND_GC` defined still something you care about?

jkotas replied:
https://github.com/dotnet/runtime/pull/91445 should fix the build with undefined `BACKGROUND_GC`

jkotas reviewed (approved):
Thank you!
